src/Components/Components/test/NavigationManagerTest.cs

```diff
@@ -208,6 +215,356 @@ public void GetUriWithQueryParameters_ThrowsWhenAnyParameterNameIsEmpty(string b
+        var navigation2 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir2", null, false);
+
+        await tcs.Task.WaitAsync(Timeout);
+
+        // Assert
+        Assert.True(navigation1.IsCompletedSuccessfully);
+        Assert.False(navigation1.Result);
+
+        Assert.True(navigation2.IsCompletedSuccessfully);
+        Assert.True(navigation2.Result);
+
+        async ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            try
+            {
+                await Task.Delay(System.Threading.Timeout.Infinite, context.CancellationToken);
+            }
+            catch (TaskCanceledException e)
+            {
+                if (e.CancellationToken == context.CancellationToken)
+                {
+                    tcs.SetResult();
+                }
+
+                throw;
+            }
+        };
+    }
+
+    [Fact]
+    public async Task LocationChangingHandlers_AreCanceledBySuccessiveNavigations_WhenMultipleHandlersAreRegistered()
+    {
+        // Arrange
+        var baseUri = "scheme://host/";
+        var navigationManager = new TestNavigationManager(baseUri);
+        var canceledHandlerTaskIds = new HashSet<string>();
+        var initialHandlerCount = 3;
+        var expectedCanceledHandlerCount = 6; // 3 handlers canceled 2 times
+        var canceledHandlerCount = 0;
+        var completedHandlerCount = 0;
+        var tcs = new TaskCompletionSource();
+
+        // Act
+        for (var i = 0; i < initialHandlerCount; i++)
+        {
+            navigationManager.AddLocationChangingHandler(HandleLocationChanging);
+        }
+
+        // These two navigations get canceled
+        var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);
+        var navigation2 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir2", null, false);
+
+        for (var i = 0; i < initialHandlerCount; i++)
+        {
+            navigationManager.RemoveLocationChangingHandler(HandleLocationChanging);
+        }
+
+        // This navigation continues without getting canceled
+        var navigation3 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir3", null, false);
+
+        await tcs.Task.WaitAsync(Timeout);
+
+        // Assert
+        Assert.True(navigation1.IsCompletedSuccessfully);
+        Assert.False(navigation1.Result);
+
+        Assert.True(navigation2.IsCompletedSuccessfully);
+        Assert.False(navigation2.Result);
+
+        Assert.True(navigation3.IsCompletedSuccessfully);
+        Assert.True(navigation3.Result);
+
+        Assert.Equal(expectedCanceledHandlerCount, canceledHandlerCount);
+        Assert.Equal(0, completedHandlerCount);
+
+        async ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            try
+            {
+                await Task.Delay(System.Threading.Timeout.Infinite, context.CancellationToken);
```

Comment:
You can do this with a TaskCompletionSource too.

```
var tcs = new TaskCompletionSource();
ct.Register(() => tcs.SetCanceled());
```

and inside this method you do await tcs.Task;

Reply:
I think the problem is that if the `CancellationToken` gets canceled before the location changing callback executes, then the TCS will never get canceled (and this actually happens in the test because the overlapping navigations are fired off synchronously). Do you think it's fine to leave as-is?

Reply:
Doesn't it work if you do `context.CancellationToken.ThrowIfCancelled()` before registering the cancellation?

Reply:
That said, I am somewhat ok leaving it as it is.